### DISCUSSION TITLE
Predefining timers during compile time?

### DISCUSSION THREAD

**Discussion by Boneman69** [arendst/Tasmota#13021] [Ideas]
Hello,

I'd really like to have the ability to define some 'standard' timers during compile time in the same way provided for rules.
We often use Sonoffs for standard purposes which all need almost the same timer settings - would be fine having the option of compiling tasmota with these timers already set...

If there is already a solution - let me know please...

Thanks in advance, Michael

**Comment by sfromis** [comment]
You can have commands in the binary, executed after first boot, using `#define USER_BACKLOG`, via rules in `#define USER_RULEn` (1, 2 or 3), or if you have a template in the build, that allows a `"CMND"` key.

**Reply by Boneman69** [reply]
Thanks for the reply, I'll give it a try.
The question is if these commands are executed only the very first boot or every time the device starts... I'll see.

**Reply by barbudor** [reply]
USER_RULEn are part of the defautlt settings of the binary. So they will be automatically configured once when the binary is flashed on a blank ESP without any settings. If you upgrade an already configured device with a binary having USER_RULEn, they will not be taken into account unless you fully erase the settings (one of the `reset` commands) to force a reconfigure with default.

USER_BACKLOG in the template will be applied the 1st time the template is applied, either from the web interface, from a MQTT `Template` command or from a initial configuration as above.

**Reply by sfromis** [reply]
To be clear, `USER_BACKLOG` works the same as `USER_RULEn`, executed only for first configuration (including after a reset to "factory settings"). What is in the template is `"CMND"` which is indeed executed only when the template is applied (including if `USER_TEMPLATE` is used), and is not kept in the saved template string.

**Reply by barbudor** [reply]
Oups, I mixed up USER_BACKLOG and template's CMND

**Reply by Boneman69** [reply]
I'm testing the solution with 'USER_BACKLOG'. I haven't found any information how long the string is allowed to be - have to try; 3 pairs of timers should be enough, hope they fit in...

**Reply by Boneman69** [reply]
Hmm, seems there are only 2 timers set because the string gets too long. 2 timers take a string length of about ~225 chars. Would have been fine to set 6 timers at once...
If there was the option of loading config files from a (predefined) website it would be possible to prepare configurations for special purposes. 
I'll use only two timers for now, maybe I could figure out to extend the 'restore config' menu...

**Reply by sfromis** [reply]
As already mentioned, you can also put such initialization into the 3 `USER_RULEn` strings (could be boot triggered rules) and into the `USER_TEMPLATE` string. Assuming much similarity in the timer JSON strings, setting them via rules could allow for a more compact form. I'm also doing setup via a MQTT based automation publishing commands. If you have 4MB of flash, you can also have an `autoexec.bat` file with commands.

**Reply by Boneman69** [reply]
I have tried to define the first timer as needed and then copy it and changing only the different values (time and action). But this seems not to work with backlog - when the first timer has been copied some values get lost (days selected and action for example)...
As the target device is a simple Sonoff basic I think there's not enough RAM to use your solution.

**Reply by Boneman69** [reply]
Ok, got it working as intended. Didn't realize I additionally had to reset the device after updating the firmware.
Now there are 6 timers set to default values as needed.

Thanks for help!

**Comment by barbudor** [comment]
@Boneman69 you gave me a idea for what I believe could be a nice feature. Almost coded, missing a few tests and will be good for PR... 😂

**Reply by Boneman69** [reply]
Hmm... I'm really curious about this :-)

**Reply by sfromis** [reply]
#13034
